### DISCUSSION TITLE
Stateful Testing in Python. Unable to satisfy assumptions of hypothesis run_state_machine.

### DISCUSSION THREAD

**Discussion by chaitubits** [schemathesis/schemathesis#1212] [Q&A]
Hi,

I am trying to get some stateful tests executed as part of a series of tests using Pytest. 
I keep getting an error like this:  Unable to satisfy assumptions of hypothesis run_state_machine.

I did update the version of my schemathesis to 3.9.0 which claims to have fixed such an error : [#965 ](url)

The schema I am using is as follows:

```
{
./v1/users/{id}:
    put:
      summary: REST API PUT Method
      description: Create or update a user.
      security:
        - xxxxxxxxxxxxx
      requestBody:
        description: Request Body with user data
        x-name: body
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/user-data'
      parameters:
        - in: path
          name: id
          required: true
          description: xxxxxxxxxxxxxxxxx
          schema:
            type: string
      responses:
        '200':
          description: 'Success'
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/response-content'
          links:
            GetUserById:
              operationId: getUser  # The target operation
              parameters:
                id: '$response.body#/id'
            DeleteUserById:
              operationId: deleteUser  # The target operation
              parameters:
                id: '$response.body#/id'
        '400':
          description: 'Invalid Request Parameters'
          content: {}
        '401':
          description: 'Unauthorized'
          content: {}
        '404':
          description: 'Item not found'
          content: {}
        '405':
          description: 'Method Not Allowed'
          content: {}
        '502':
          description: 'Bad Gateway'
          content: {}
    get:
      summary: REST API GET Method
      operationId: getUser
      description: Returns user info
      security:
        - xxxxxxxxxxxxxxxxxxxxx
      parameters:
        - in: path
          name: id
          required: true
          description: Configuration ID
          schema:
            type: string
      responses:
        '200':
          description: 'Success'
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/user-data' 
}

```

And my python code:

```python
@pytest.fixture
def state_machine(schema, read_token, write_token):
    class APIWorkflow(schema.as_state_machine()):
        def before_call(self, case):
            case.headers = {}
            if case.method == "GET":
                case.headers["Authorization"] = f"Bearer {read_token}"
            else:
                case.headers["Authorization"] = f"Bearer {write_token}"

        def validate_response(self, response, case):
            if case.method == "GET" and case.path == "/openapi.json":
                case.validate_response(response, checks=(my_check,))

    return APIWorkflow

def test_statefully(state_machine):
    run_state_machine_as_test(
        state_machine,
        settings=settings(
            max_examples=1,
            deadline=None,
            verbosity=Verbosity.verbose,
            suppress_health_check=(HealthCheck.all()),
            stateful_step_count=5,
        ),
    )

```

Any help is appreciated. Thanks!

P.S: I tried to make the schema and code look better, but wasn't able to due to my bad skills, hope that can be overlooked ..

**Comment by chaitubits** [comment]
Any help ?

**Reply by Stranger6667** [reply]
Hi, sorry for not answering for so long. I'll take a look at it tomorrow

**Reply by Stranger6667** [reply]
> I keep getting an error like this: Unable to satisfy assumptions of hypothesis run_state_machine.
> I did update the version of my schemathesis to 3.9.0 which claims to have fixed such an error : #965

It seems like it might be caused by something else. Could you please share the schema with used `components`  (`user-data` and `response-content`)? Unfortunately, I wasn't able to reproduce it locally with the schema above, where I replaced missing components with `{}` to make it work.

Also, how do you load schema? Is it with `schemathesis.from_dict` or some other loader? It will be nice to see all the arguments passed to it.

Also, does it send any requests to the tested app? Or just reports this error?

A probable reason might be that the actual payload is too complex for Schemathesis to generate (e.g. some deeply nested schemas) or just unsatisfiable ones (often happens with `oneOf`). For example, if I change the `id` schemas to `{"pattern": r"\d+\w", "maxLength": 1, "type": "string"}` which is unsatisfiable (and we can't check it at the moment), then I get `hypothesis.errors.Unsatisfiable: Unable to satisfy assumptions of hypothesis run_state_machine.`. So, having the complete schema might uncover the underlying reason.

In any case, Schemathesis could give better error messages (https://github.com/schemathesis/schemathesis/issues/1083)

**Reply by chaitubits** [reply]
No problem, Thanks much for the response.
I am loading the schema from a json file located at {base_url}/openapi.json.

I will try to get more details as requested for the schema.
